### PR TITLE
Reintroduce a package for Razor runtime compilation

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/4947

### CODE REVIEW

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModelFactory.cs [hunk 1]
```diff
@@ -7,18 +7,27 @@
 using Microsoft.AspNetCore.Mvc.Razor;
 using Microsoft.AspNetCore.Mvc.RazorPages;
 using Microsoft.Extensions.Logging;
-using Microsoft.Extensions.Primitives;
 
 namespace Microsoft.AspNetCore.Mvc.ApplicationModels
 {
     internal class PageRouteModelFactory
     {
+        private static readonly Action<ILogger, string, Exception> _unsupportedAreaPath;
```
Comment (dougbu) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModelFactory.cs]: Why is this moving from `PageLoggerExtensions`?

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/FileProviderRazorProjectFileSystemTest.cs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                    Assert.Equal("/", file.BasePath);
+                },
+                file =>
+                {
+                    Assert.Equal("/File3.cshtml", file.FilePath);
+                    Assert.Equal("/", file.BasePath);
+                });
+        }
+
+        //[Fact]
```
Comment (davidfowl) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/FileProviderRazorProjectFileSystemTest.cs]: WIP?

**Review by dougbu — COMMENTED** [review]
Please help us review this PR:
- What files need to be reviewed in detail i.e. what files are not just being restored from earlier commits?
- Is any part of this separate i.e. something to consider as a smaller PR?

**Comment by pranavkm** [comment]
Let's do the PR in person. It'd be easier to diff the changes against the 2.2 branch of MVC and focus on the things that are new \ different.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/ViewEngineTests.cs [hunk 3]
```diff
@@ -252,30 +252,6 @@ public async Task RazorViewEngine_RendersPartialViews(string actionName, string
...
 
-        [Fact(Skip = "https://github.com/aspnet/Mvc/issues/8754")]
-        public Task RazorViewEngine_RendersViewsFromEmbeddedFileProvider_WhenLookedupByName()
-            => RazorViewEngine_RendersIndexViewsFromEmbeddedFileProvider("/EmbeddedViews/LookupByName");
-
-        [Fact(Skip = "https://github.com/aspnet/Mvc/issues/8754")]
-        public Task RazorViewEngine_RendersViewsFromEmbeddedFileProvider_WhenLookedupByPath()
-            => RazorViewEngine_RendersIndexViewsFromEmbeddedFileProvider("/EmbeddedViews/LookupByPath");
-
-        private async Task RazorViewEngine_RendersIndexViewsFromEmbeddedFileProvider(string requestPath)
```
Comment (dougbu) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/ViewEngineTests.cs]: I see you removed the web site's support for this. Would have thought embedded views would work if runtime complication is enabled. Is the combination tested elsewhere?

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/CSharpCompilerTest.cs [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+            // Act & Assert
+            var parseOptions = compiler.ParseOptions;
+            Assert.Contains("SOME_TEST_DEFINE", parseOptions.PreprocessorSymbolNames);
+        }
+
+        [Fact]
+        public void Constructor_ConfiguresLanguageVersion()
+        {
+            // Arrange
+            var dependencyContextOptions = new DependencyContextCompilationOptions(
```
Comment (dougbu) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/CSharpCompilerTest.cs]: This is repeated many times. Suggest using a helper to make it clear where the options are different.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/ChecksumValidatorTest.cs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            Assert.False(result);
+        }
+
+        [Fact]
+        public void IsRecompilationSupported_NoPrimaryChecksum_ReturnsFalse()
+        {
+            // Arrange
+            var item = new TestRazorCompiledItem(typeof(string), "mvc.1.0.view", "/Views/Home/Index.cstml", new object[]
+            {
+                new RazorSourceChecksumAttribute("SHA1", GetChecksum("some import"), "/Views/Home/_ViewImports.cstml"),
```
Comment (dougbu) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/ChecksumValidatorTest.cs]: This isn't about checksums. It's about primary content. Suggest renaming the test.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/DependencyInjection/RazorRuntimeCompilationMvcCoreBuilderExtensionsTest.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+            // Arrange
+            var services = new ServiceCollection()
+                .AddSingleton<IViewCompilerProvider, DefaultViewCompilerProvider>();
+
+            // Act
+            RazorRuntimeCompilationMvcCoreBuilderExtensions.AddServices(services);
+
+            // Assert
+            var serviceDescriptor = Assert.Single(services, service => service.ServiceType == typeof(IViewCompilerProvider));
+            Assert.Equal(typeof(RuntimeViewCompilerProvider), serviceDescriptor.ImplementationType);
```
Comment (dougbu) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/DependencyInjection/RazorRuntimeCompilationMvcCoreBuilderExtensionsTest.cs]: Minor but `Assert.IsType<RuntimeViewCompilerProvider>(...)` is easier to read

**Review by dougbu — APPROVED** [review]
There's a lot here and I may have missed a few details. But, there's more than enough tests to catch most real problems.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.csproj [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <Description>Runtime compilation support for Razor views and Razor Pages in ASP.NET Core MVC.</Description>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <NoWarn>$(NoWarn);CS1591</NoWarn>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;aspnetcoremvc;razor</PackageTags>
```
Comment (natemcmaster) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.csproj]:
```suggestion
    <PackageTags>aspnetcore;aspnetcoremvc;razor</PackageTags>
    <IsProductPackage>true</IsProductPackage>
```

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/ViewEngineTests.cs (hunk 3 above)
Comment (pranavkm) [reply src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/ViewEngineTests.cs]: The functional tests use an in-memory file provider which acts pretty much like the embedded one. These specific tests were used to verify path manipulation for files not coming from the file provider which is covered by default due to Razor Sdk compiled views.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/DependencyInjection/RazorRuntimeCompilationMvcCoreBuilderExtensionsTest.cs (hunk 6 above)
Comment (pranavkm) [reply src/Mvc/test/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.Test/DependencyInjection/RazorRuntimeCompilationMvcCoreBuilderExtensionsTest.cs]: That only works if you're testing an instance e.g. `Assert.IsType<RuntimeViewCompilerProvider>(value);`, not when you're just asserting two instances of `Type`

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.csproj [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    <IsProductPackage>true</IsProductPackage>
```
Comment (natemcmaster) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.csproj]:
```suggestion
    <IsShippingPackage>true</IsShippingPackage>
```

I renamed this property recently to align with Arcade naming.

**Review by natemcmaster — APPROVED** [review]
Only reviewed build infrastructure changes. One minor fix in .csproj is required, but otherwise looks fine to me.

**Comment by pranavkm** [comment]
🆙 📅 with naming change

**Pull request merged**